iceoryx_posh/doc/chunk_header.md

```diff
@@ -0,0 +1,218 @@
+|---------------------------------------------------------------------------->|
+```
+
+### Payload Offset Calculation
+
+1. No custom header and payload alignment doesn't exceed the `ChunkHeader` alignment
+
+```
+payloadOffset = sizeof(ChunkHeader);
+```
+
+2. No custom header and payload alignment exceeds the `ChunkHeader` alignment, which means the payload is either adjacent to the `ChunkHeader` or there is a padding with at least the size of `ChunkHeader` in front of the payload and therefore enough space to store the `back-offset`
+
+```
+headerEndAddress = addressof(chunkHeader) + sizeof(chunkHeader);
+payloadAddress = align(headerEndAddress, payloadAlignment);
+payloadOffset = payloadAddress - addressof(chunkHeader);
+```
+
+3. Custom header is used
+
+```
+headerEndAddress = addressof(chunkHeader) + sizeof(chunkHeader) + sizeof(customHeader);
+potentialBackOffsetAddress = align(headerEndAddress, alignof(payloadOffset));
+potentialPayloadAddress = potentialBackOffsetAddress + sizeof(payloadOffset);
+payloadAddress = align(potentialPayloadAddress, payloadAlignment);
+payloadOffset = payloadAddress - addressof(chunkHeader);
+```
+
+### Required Chunk Size Calculation
+
+In order to fit the custom header and the payload into he chunk, a worst case calculation has to be done. We can assume that a chunk is aligned to 32 bytes, which is also the alignment of the `ChunkHeader`.
+
+1. No custom header and payload alignment doesn't exceed the `ChunkHeader` alignment
+
+```
+chunkSize = sizeof(chunkHeader) + payloadSize;
+```
+
+2. No custom header and payload alignment exceeds the `ChunkHeader` alignment
+
+Worst case scenario is when a part of the `ChunkHeader` crosses the payload alignment boundary, so that the payload must be aligned to the next boundary.
+```
+sizeLeftOfThePayloandAlignmentBoundary = sizeof(chunkHeader) - alignof(chunkHeader);
+chunkSize = sizeLeftOfThePayloandAlignmentBoundary + payloadAlignment + payloadSize;
+```
+
+3. Custom header is used
+
+Similar to case 2, but in this case it is the `back-offset` which might cross the payload alignment boundary.
+```
+headerSize = sizeof(chunkHeader) + sizeof(customHeader)
+sizeLeftOfThePayloandAlignmentBoundary = align(headerSize, alignof(payloadOffset));
+maxAlignment = max(alignof(payloadOffset), payloadAlignment);
+chunkSize = sizeLeftOfThePayloandAlignmentBoundary + maxAlignment + payloadSize;
+```
+
+### Accessing Chunk Header Extension
+
+The `ChunkHeader` has a template method to get custom header. There is a risk to use this wrong by accident, but there is currently no better solution for this.
+
+Since the custom header is always adjacent to the `ChunkHeader`, the formula to get the custom header is:
+```
+customHeader = addressOf(chunkHeader) + sizeof(chunkHeader);
+```
+
+### ChunkHeader Methods
+
+- `void* payload()` returns a pointer to the payload
+- `template <typename T> T* customHeader()` returns a pointer to the custom header
```

Comment:
what is returned if there is no custom header?

Reply:
I suppose we could use SFINAE (`enable_if` meta programming magic) to prevent the user calling this method in case there is no custom header.

Reply:
if we store a flag, a nullptr. Maybe an `optional<T*>` would be better